packages/firestore/src/model/mutation.ts

```diff
@@ -302,17 +276,25 @@ export abstract class Mutation {
     localWriteTime: Timestamp
   ): MaybeDocument | null;
 
-  abstract isEqual(other: Mutation): boolean;
-
   /**
-   * If applicable, returns the field mask for this mutation. Fields that are
-   * not included in this field mask are not modified when this mutation is
-   * applied. Mutations that replace the entire document return 'null'.
+   * If applicable, returns the base value to persist with this mutation. If a
+   * base value is provided, the mutation is always applied to this base value,
+   * even if document has already beenupdated.
+   *
+   * The base value is a sparse object that consists of only the document
+   * fields for which this mutation contains a non-idempotent transformation
+   * (e.g. a numeric increment). The provided alue guarantees consistent
+   * behavior for non-idempotent transforms and allow us to return the same
+   * latency-compensated value even if the backend has already applied the
+   * mutation. The base value is null for idempotent mutations, as they can be
+   * re-played even if the backend has already applied them.
+   *
+   * @return a base value to store along with the mutation, or null for
+   * idempotent mutations.
    */
-  abstract get fieldMask(): FieldMask | null;
+  abstract extractBaseValue(maybeDoc: MaybeDocument | null): ObjectValue | null;
```

Comment:
As a reader who don't have enough knowledge, `BaseValue` is a little vague. I could not understand it until after reading the comments. I feel like maybe we should try to come up with a better name, my attempt is `computeBaseValueToApplyMutation`... a bit wordy.

Reply:
I can't come up with a better term that is still concise. For better or for worse, the "base" concept is already used in our serialization (https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/local/indexeddb_schema.ts#L429) and may just be one of the core concepts that someone has to ramp up on before they can understand how our mutations are applied. I agree that it would be nicer if the name alone conveyed what this value is used for, but I can't come up with anything short enough to win me over.